### ISSUE TITLE
Some questions about the 'test_model' function

### ISSUE THREAD

**Issue by bbjy** [Zziwei/Joint-Collaborative-Autoencoder#1]
Hi, thank you for your work! I have some questions about it.

1.  You don't use the validation set, do you? So the experiment only has two parts: training and testing, is it?
2.   Why the 'self.input_P_cor' and 'self.input_N_cor' in 'test_model' is '[0,0]', rather than feed it the same as ' train_model'? 
 
I am looking forward to your reply! Thank you so much!

**Comment by Zziwei** [comment]
Thanks for your interest.

1. For the vali set, you should modify the test.py code by using '.train()' function. (such as yelp.train(), which will return the training and validation sets).

2. Because for inferring, we do not need negative sampling, just need to feed the whole interaction matrix to it and reconstruct it.

**Comment by Zziwei** [comment]
> Hi, thank you for your work! I have some questions about it.
> 
> 1. You don't use the validation set, do you? So the experiment only has two parts: training and testing, is it?
> 2. Why the 'self.input_P_cor' and 'self.input_N_cor' in 'test_model' is '[0,0]', rather than feed it the same as ' train_model'?
> 
> I am looking forward to your reply! Thank you so much!

Thanks for your interest.

For the vali set, you should modify the test.py code by using '.train()' function. (such as yelp.train(), which will return the training and validation sets).

Because for inferring, we do not need negative sampling, just need to feed the whole interaction matrix to it and reconstruct it.

**Comment by bbjy** [comment]
> > Hi, thank you for your work! I have some questions about it.
> > 
> > 1. You don't use the validation set, do you? So the experiment only has two parts: training and testing, is it?
> > 2. Why the 'self.input_P_cor' and 'self.input_N_cor' in 'test_model' is '[0,0]', rather than feed it the same as ' train_model'?
> > 
> > I am looking forward to your reply! Thank you so much!
> 
> Thanks for your interest.
> 
> For the vali set, you should modify the test.py code by using '.train()' function. (such as yelp.train(), which will return the training and validation sets).
> 
> Because for inferring, we do not need negative sampling, just need to feed the whole interaction matrix to it and reconstruct it.

Thanks for your reply.

I know that when I modify the code to '.train()' , it will return the training and validation sets. But, when I want to test, it will run 'train_model'  again as the code in 'run' function is written. In my view, the process of running the program should be 'training', 'validation' ,and then 'testing'. Is there any mistake in my understanding?

For the second problem,  thanks for your explanation it helps me understand that the calculation method of 'test cost'  is different from the 'training cost'.

**Comment by Zziwei** [comment]
> > > Hi, thank you for your work! I have some questions about it.
> > > 
> > > 1. You don't use the validation set, do you? So the experiment only has two parts: training and testing, is it?
> > > 2. Why the 'self.input_P_cor' and 'self.input_N_cor' in 'test_model' is '[0,0]', rather than feed it the same as ' train_model'?
> > > 
> > > I am looking forward to your reply! Thank you so much!
> > 
> > 
> > Thanks for your interest.
> > For the vali set, you should modify the test.py code by using '.train()' function. (such as yelp.train(), which will return the training and validation sets).
> > Because for inferring, we do not need negative sampling, just need to feed the whole interaction matrix to it and reconstruct it.
> 
> Thanks for your reply.
> 
> I know that when I modify the code to '.train()' , it will return the training and validation sets. But, when I want to test, it will run 'train_model' again as the code in 'run' function is written. In my view, the process of running the program should be 'training', 'validation' ,and then 'testing'. Is there any mistake in my understanding?
> 
> For the second problem, thanks for your explanation it helps me understand that the calculation method of 'test cost' is different from the 'training cost'.

I first use the validation set to tune the hyperparameters, then use the test set to do the final evaluation. When doing the test, I retrained the model using the best hyperparameters based on vali set.

**Comment by bbjy** [comment]
> > > > Hi, thank you for your work! I have some questions about it.
> > > > 
> > > > 1. You don't use the validation set, do you? So the experiment only has two parts: training and testing, is it?
> > > > 2. Why the 'self.input_P_cor' and 'self.input_N_cor' in 'test_model' is '[0,0]', rather than feed it the same as ' train_model'?
> > > > 
> > > > I am looking forward to your reply! Thank you so much!
> > > 
> > > 
> > > Thanks for your interest.
> > > For the vali set, you should modify the test.py code by using '.train()' function. (such as yelp.train(), which will return the training and validation sets).
> > > Because for inferring, we do not need negative sampling, just need to feed the whole interaction matrix to it and reconstruct it.
> > 
> > 
> > Thanks for your reply.
> > I know that when I modify the code to '.train()' , it will return the training and validation sets. But, when I want to test, it will run 'train_model' again as the code in 'run' function is written. In my view, the process of running the program should be 'training', 'validation' ,and then 'testing'. Is there any mistake in my understanding?
> > For the second problem, thanks for your explanation it helps me understand that the calculation method of 'test cost' is different from the 'training cost'.
> 
> I first use the validation set to tune the hyperparameters, then use the test set to do the final evaluation. When doing the test, I retrained the model using the best hyperparameters based on vali set.

I got it. Thank you!
